# Problem Statement - San Francisco Crime Classification
Predict the category of crimes that occurred in the city by the bay 
To examine the specific problem, we will apply a full Data Science life cycle composed of the following steps:

1. Data Wrangling : to audit the quality of the data and perform all the necessary actions to clean the dataset.
2. Data Exploration (EDA) : for understanding the predictors and create intuition on the data.
3. Feature Engineering : to create additional predictors from the existing which are potentially more insightful.
4. Data Normalization and Data Transformation : for preparing the dataset for learning algorithms (if needed).
6. Training / Testing data creation : to evaluate the performance of our models and fine-tune their hyperparameters.
7. Model selection and evaluation. This will be the final goal; creating a model that predicts the probability of each type of crime based on the location and the date.

# Data
This dataset contains incidents derived from SFPD Crime Incident Reporting system. The data ranges from 1/1/2003 to 5/13/2015. The training set and test set rotate every week, meaning week 1,3,5,7... belong to test set, week 2,4,6,8 belong to training set. 

Data fields
1. Dates - timestamp of the crime incident
2. Category - category of the crime incident (only in train.csv). This is the target variable you are going to predict.
3. Descript - detailed description of the crime incident (only in train.csv)
4. DayOfWeek - the day of the week
5. PdDistrict - name of the Police Department District
6. Resolution - how the crime incident was resolved (only in train.csv)
7. Address - the approximate street address of the crime incident 
8. X - Longitude
9. Y - Latitude

## Analysis

In [2]:
import pandas as pd

In [9]:
train = pd.read_csv('input/train.csv', parse_dates=['Dates'])
train.head()

Dates        Category                      Descript  \
0 2015-05-13 23:53:00        WARRANTS                WARRANT ARREST   
1 2015-05-13 23:53:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
2 2015-05-13 23:33:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
3 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   
4 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   

   DayOfWeek PdDistrict      Resolution                    Address  \
0  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
1  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
2  Wednesday   NORTHERN  ARREST, BOOKED  VANNESS AV / GREENWICH ST   
3  Wednesday   NORTHERN            NONE   1500 Block of LOMBARD ST   
4  Wednesday       PARK            NONE  100 Block of BRODERICK ST   

            X          Y  
0 -122.425892  37.774599  
1 -122.425892  37.774599  
2 -122.424363  37.800414  
3 -122.426995  37.800873  
4 -122.438738  37.771541

In [8]:
test = pd.read_csv('input/test.csv', parse_dates=['Dates'], index_col='Id')
test.head()

Dates DayOfWeek PdDistrict                   Address  \
Id                                                                      
0  2015-05-10 23:59:00    Sunday    BAYVIEW   2000 Block of THOMAS AV   
1  2015-05-10 23:51:00    Sunday    BAYVIEW        3RD ST / REVERE AV   
2  2015-05-10 23:50:00    Sunday   NORTHERN    2000 Block of GOUGH ST   
3  2015-05-10 23:45:00    Sunday  INGLESIDE  4700 Block of MISSION ST   
4  2015-05-10 23:45:00    Sunday  INGLESIDE  4700 Block of MISSION ST   

             X          Y  
Id                         
0  -122.399588  37.735051  
1  -122.391523  37.732432  
2  -122.426002  37.792212  
3  -122.437394  37.721412  
4  -122.437394  37.721412

In [12]:
print('First Date: ', str(train.Dates.describe()['first']))
print('Last Date: ', str(train.Dates.describe()['last']))
print('Train Data Shape', train.shape)

First Date:  2003-01-06 00:01:00
Last Date:  2015-05-13 23:53:00
Train Data Shape (878049, 9)


In [13]:
train.head()

Dates        Category                      Descript  \
0 2015-05-13 23:53:00        WARRANTS                WARRANT ARREST   
1 2015-05-13 23:53:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
2 2015-05-13 23:33:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
3 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   
4 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   

   DayOfWeek PdDistrict      Resolution                    Address  \
0  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
1  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
2  Wednesday   NORTHERN  ARREST, BOOKED  VANNESS AV / GREENWICH ST   
3  Wednesday   NORTHERN            NONE   1500 Block of LOMBARD ST   
4  Wednesday       PARK            NONE  100 Block of BRODERICK ST   

            X          Y  
0 -122.425892  37.774599  
1 -122.425892  37.774599  
2 -122.424363  37.800414  
3 -122.426995  37.800873  
4 -122.438738  37.771541

In [14]:
train.dtypes

Dates         datetime64[ns]
Category              object
Descript              object
DayOfWeek             object
PdDistrict            object
Resolution            object
Address               object
X                    float64
Y                    float64
dtype: object

The dataset contains a lot of 'object' variables (aka strings) that we will need to encode.

In [15]:
train.duplicated().sum()

2323

It also contains 2323 duplicates that we must remove.
We will also evaluate the position of the data points using the coordinates.